tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+            unsafe { EntryState::from_repr(self.0.load(ordering)) }
+        }
+
+        pub(super) fn set(&self, s: EntryState, ordering: Ordering) {
+            self.0.store(s.to_repr(), ordering)
         }
     }
 
-    /// Called by `Timer` to load the current value of `state` for processing
-    pub(crate) fn load_state(&self) -> Option<u64> {
-        let state = self.state.load(SeqCst);
+    pub(super) struct StateRef(std::sync::Arc<AtomicU8>);
+
+    impl StateRef {
+        pub(super) fn get(&self, ordering: Ordering) -> EntryState {
+            unsafe { EntryState::from_repr(self.0.load(ordering)) }
+        }
 
-        if is_elapsed(state) {
-            None
-        } else {
-            Some(state)
+        pub(super) fn set(&self, s: EntryState, ordering: Ordering) {
+            self.0.store(s.to_repr(), ordering)
         }
     }
+}
 
-    pub(crate) fn is_elapsed(&self) -> bool {
-        let state = self.state.load(SeqCst);
-        is_elapsed(state)
+use state_cell::StateCell;
+
+impl std::fmt::Debug for StateCell {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "StateCell({:?})", self.get(Ordering::SeqCst))
     }
+}
+
+/// A timer entry.
+///
+/// This is the handle to a timer that is controlled by the requester of the
+/// timer. As this participates in intrusive data structures, it must be pinned
+/// before polling.
+#[derive(Debug)]
+pub(super) struct TimerEntry<TS: TimeSource> {
+    /// Arc reference to the driver. We can only free the driver after
+    /// deregistering everything from their respective timer wheels.
+    driver: InternalHandle<TS>,
+    /// Shared inner structure; this is part of an intrusive linked list, and
+    /// therefore other references can exist to it while mutable references to
+    /// Entry exist.
+    ///
+    /// This is manipulated only under the inner mutex. TODO: Can we use loom
+    /// cells for this?
+    inner: StdUnsafeCell<TimerShared>,
+}
 
-    pub(crate) fn fire(&self, when: u64) {
-        let mut curr = self.state.load(SeqCst);
+unsafe impl<TS: TimeSource> Send for TimerEntry<TS> {}
+unsafe impl<TS: TimeSource> Sync for TimerEntry<TS> {}
 
-        loop {
-            if is_elapsed(curr) || curr > when {
-                return;
-            }
+/// An EntryHandle is the (non-enforced) "unique" pointer from the driver to the
+/// timer entry. Generally, at most one EntryHandle exists for a timer at a time
+/// (enforced by the timer state machine).
+///
+/// SAFETY: An EntryHandle is essentially a raw pointer, and the usual caveats
+/// of pointer safety apply. In particular, EntryHandle does not itself enforce
+/// that the timer does still exist; however, normally an EntryHandle is created
+/// immediately before registering the timer, and is consumed when firing the
+/// timer, to help minimize mistakes. Still, because EntryHandle cannot enforce
+/// memory safety, all operations are unsafe.
```

Comment:
nit: looks like the name of the type changed?
```suggestion
/// An TimerHandle is the (non-enforced) "unique" pointer from the driver to the
/// timer entry. Generally, at most one TimerHandle exists for a timer at a time
/// (enforced by the timer state machine).
///
/// SAFETY: An TimerHandle is essentially a raw pointer, and the usual caveats
/// of pointer safety apply. In particular, TimerHandle does not itself enforce
/// that the timer does still exist; however, normally an TimerHandle is created
/// immediately before registering the timer, and is consumed when firing the
/// timer, to help minimize mistakes. Still, because TimerHandle cannot enforce
/// memory safety, all operations are unsafe.
```

Reply:
yep, looks like I missed a spot